File: motion_primitives_forward_controller/src/fk_client.cpp
// Copyright (c) 2025, b»robotized
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Authors: Mathias Fuhrer

#include "motion_primitives_forward_controller/fk_client.hpp"

FKClient::FKClient(const std::string & node_name) : Node(node_name)
{
  fk_client_ = this->create_client<moveit_msgs::srv::GetPositionFK>("/compute_fk");

  while (!fk_client_->wait_for_service(std::chrono::seconds(1)))
  {
    RCLCPP_INFO(this->get_logger(), "Waiting for /compute_fk service...");
  }
}

geometry_msgs::msg::Pose FKClient::computeFK(
  const std::vector<std::string> & joint_names, const std::vector<double> & joint_positions,
  const std::string & from_frame, const std::string & to_link)
{
  auto request = std::make_shared<moveit_msgs::srv::GetPositionFK::Request>();
  request->fk_link_names.push_back(to_link);

  sensor_msgs::msg::JointState joint_state;
  joint_state.name = joint_names;
  joint_state.position = joint_positions;
  request->robot_state.joint_state = joint_state;
  request->header.frame_id = from_frame;

  auto future = fk_client_->async_send_request(request);

  if (
    rclcpp::spin_until_future_complete(this->get_node_base_interface(), future) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    auto response = future.get();
    if (!response->pose_stamped.empty())
    {
      return response->pose_stamped.front().pose;
    }
    else
    {
      throw std::runtime_error("Empty response received from FK service.");
    }
  }
  else
  {
    throw std::runtime_error("Error calling FK service.");
  }
}

File: motion_primitives_forward_controller/motion_primitives_forward_controller/fk_client.py
#!/usr/bin/env python3

# Copyright (c) 2025, b»robotized
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Authors: Mathias Fuhrer

import rclpy
from rclpy.node import Node
from moveit_msgs.srv import GetPositionFK


class FKClient(Node):
    def __init__(self):
        rclpy.init()
        super().__init__("fk_client")
        self.client = self.create_client(GetPositionFK, "/compute_fk")

        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Waiting for /compute_fk service...")

    def compute_fk(self, joint_names, joint_positions, from_frame="base", to_link="tool0"):
        request = GetPositionFK.Request()
        request.header.frame_id = from_frame
        request.fk_link_names = [to_link]
        request.robot_state.joint_state.name = joint_names
        request.robot_state.joint_state.position = joint_positions

        future = self.client.call_async(request)
        rclpy.spin_until_future_complete(self, future, timeout_sec=3.0)

        if future.done():
            result = future.result()
            if result and result.error_code.val == 1:
                return result.pose_stamped[0].pose
            else:
                self.get_logger().warn(f"FK error: code={result.error_code.val}")
        else:
            self.get_logger().error("FK call timed out")

        return None

    def shutdown(self):
        self.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    fk = FKClient()
    joint_names = [
        "shoulder_pan_joint",
        "shoulder_lift_joint",
        "elbow_joint",
        "wrist_1_joint",
        "wrist_2_joint",
        "wrist_3_joint",
    ]
    pose = fk.compute_fk(joint_names, [0.0, 1.0, 0.5, 0.0, 0.0, 0.0])
    if pose:
        print(pose)
    fk.shutdown()
